### PR TITLE
UOE-9429: Add DB stats in pubmatic module

### PR DESCRIPTION
**Pull request by pm-saurabh-narkhede**
# Description

Please add change description or link to ticket, docs, etc.

# Checklist:

- [ ] PR commit list is unique (rebase/pull with the origin branch to keep master clean).
- [ ] JIRA number is added in the PR title and the commit message.
- [ ] Updated the `header-bidding` repo with appropiate commit id.
- [ ] Documented the new changes.

For Prebid upgrade, refer: https://inside.pubmatic.com:8443/confluence/display/Products/Prebid-server+upgrade

### CODE REVIEW

File: modules/pubmatic/openwrap/cache/gocache/util.go [hunk 1]
```diff
@@ -23,3 +24,12 @@ func validUPixels(upixel []adunitconfig.UniversalPixel) []adunitconfig.Universal
 	}
 	return validPixels
 }
+
+// wraps error with error msg
+func errorWrap(cErr, nErr error) error {
```
Comment (AvinashKapre) [inline modules/pubmatic/openwrap/cache/gocache/util.go:29]: Add test case for these

File: modules/pubmatic/openwrap/metrics/config/metrics.go [hunk 2]
```diff
@@ -361,6 +362,34 @@ func (me *MultiMetricsEngine) RecordVideoImpDisabledViaConnTypeStats(publisher,
...
+		thisME.RecordSendLoggerDataTime(endpoint, profile, sendTime)
+	}
+}
+
+// RecordRequestTime across all engines
+func (me *MultiMetricsEngine) RecordRequestTime(endpoint string, sendTime time.Duration) {
+	for _, thisME := range *me {
+		thisME.RecordRequestTime(endpoint, sendTime)
+	}
+}
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/metrics/config/metrics.go:384]: this is available in PBS code, we can add label there

File: modules/pubmatic/openwrap/database/mysql/slot_mapping.go [hunk 3]
```diff
@@ -26,8 +27,8 @@ func (db *mySqlDB) GetPubmaticSlotMappings(pubID int) (map[string]models.SlotMap
 
 		err := rows.Scan(&slotInfo.SlotName, &slotInfo.AdSize, &slotInfo.SiteId,
 			&slotInfo.AdTagId, &slotInfo.GId, &slotInfo.Floor)
-		if nil != err {
-			//continue
+		if err != nil {
+			glog.Errorf("pubmatic slot mappings row scan failed for %d", pubID)
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/database/mysql/slot_mapping.go:31]: profile id would be helpful here

File: modules/pubmatic/openwrap/database/mysql/partner_config.go [hunk 4]
```diff
@@ -69,7 +70,7 @@ func (db *mySqlDB) getActivePartnerConfigurations(versionID int) (map[int]map[st
 
 	// NYC_TODO: ignore close error
 	if err = rows.Err(); err != nil {
-
+		glog.Error("partner config row scan failed")
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/database/mysql/partner_config.go:73]: This is very vague log. Would be helpful to log pub, profile, etc to debug.

File: modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go [hunk 5]
```diff
@@ -2,11 +2,14 @@ package adunitconfig
 
 import (
 	"encoding/json"
+	"errors"
 
 	"github.com/prebid/openrtb/v19/openrtb2"
 	"github.com/prebid/prebid-server/openrtb_ext"
 )
 
+var ErrAdUnitUnmarshal = errors.New("unmarshal error adunitconfig")
```
Comment (pm-aadit-patil) [inline modules/pubmatic/openwrap/models/adunitconfig/adunitconfig.go:11]: suggestion : should we move all error const declarations to any dedicated file?

**Review by AvinashKapre — DISMISSED** [review]

File: modules/pubmatic/openwrap/cache/gocache/partner_config_test.go [hunk 6]
```diff
@@ -10,14 +11,17 @@ import (
...
 	"github.com/prebid/prebid-server/modules/pubmatic/openwrap/models"
 	"github.com/stretchr/testify/assert"
 )
 
 func Test_cache_GetPartnerConfigMap(t *testing.T) {
+	var queryFailCase bool
 	ctrl := gomock.NewController(t)
 	defer ctrl.Finish()
 	mockDatabase := mock_database.NewMockDatabase(ctrl)
+	mockEngine := mock.NewMockMetricsEngine(ctrl)
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/cache/gocache/partner_config_test.go:24]: move this to per test in `setup` for better UT

**Review by pm-nilesh-chate — COMMENTED** [review]
added minor comments

File: modules/pubmatic/openwrap/cache/gocache/partner_config_test.go [hunk 7]
```diff
@@ -69,24 +232,29 @@ func Test_cache_GetPartnerConfigMap(t *testing.T) {
...
+			defer ctrl.Finish()
+
 			c := &cache{
 				cache: tt.fields.cache,
 				cfg:   tt.fields.cfg,
-				db:    tt.fields.db,
 			}
+
+			c.db, c.metricEngine = tt.setup(ctrl)
+			defer ctrl.Finish()
```
Comment (pm-nilesh-chate) [inline modules/pubmatic/openwrap/cache/gocache/partner_config_test.go:251]: `defer ctrl.Finish()` called 2 times

**Review by pm-nilesh-chate — APPROVED** [review]

**Review by pm-nilesh-chate — APPROVED** [review]

**Review by AvinashKapre — APPROVED** [review]

**Pull request merged**